### PR TITLE
nspawn: add --suppress-sync=yes mode for turning sync() and friends i…

### PR DESCRIPTION
**Pull request by poettering**
…nto NOPs via seccomp

This is supposed to be used by package/image builders such as mkosi to
speed up building, since it allows us to suppress sync() inside a
container.

This does what Debian's eatmydata tool does, but for a container, and
via seccomp (instead of LD_PRELOAD).

### CODE REVIEW

File: src/nspawn/nspawn.c [hunk 1]
```diff
@@ -342,7 +343,9 @@ static int help(void) {
                "  -E --setenv=NAME[=VALUE]  Pass an environment variable to PID 1\n"
                "  -u --user=USER            Run the command under specified user or UID\n"
                "     --kill-signal=SIGNAL   Select signal to use for shutting down PID 1\n"
-               "     --notify-ready=BOOLEAN Receive notifications from the child init process\n\n"
+               "     --notify-ready=BOOLEAN Receive notifications from the child init process\n"
+               "     --suppress-sync=BOOLEAN\n"
+               "                            Suppress any form of disk data synchronization\n"
```
Comment (yuwata) [inline src/nspawn/nspawn.c]: Missing blank line, i.e., `\n` -> `\n\n`. Is this intended?

Comment (poettering) [reply src/nspawn/nspawn.c]: nope

**Comment by keszybz** [comment]
Looks great. I tested this with deboostrap, and the annoying syncs are all suppressed. I think this is good to merge, but I'll leave it open until tomorrow if others have some comments.

File: src/shared/seccomp-util.c [hunk 2]
```diff
@@ -2205,3 +2205,97 @@ int parse_syscall_and_errno(const char *in, char **name, int *error) {
...
+#endif
+
+        return any ? 0 : r;
+}
+
+int seccomp_suppress_sync(void) {
+        uint32_t arch;
+        int r;
+
+        /* This is mostly identical to SystemCallFilter=~@sync:0, but simpler to use, and seperately manageable. */
```
Comment (bluca) [inline src/shared/seccomp-util.c]: /seperately/separately/

File: src/shared/seccomp-util.c [hunk 3]
```diff
@@ -2205,3 +2205,97 @@ int parse_syscall_and_errno(const char *in, char **name, int *error) {
...
+                                continue;
+                        }
+
+                        r = seccomp_rule_add_exact(
+                                        seccomp,
+                                        SCMP_ACT_ERRNO(0), /* success → we want this to be a NOP after all */
+                                        id,
+                                        0);
+                        if (r < 0)
+                                log_debug_errno(r, "Failed to add filter for system call %s: %m", c);
```
Comment (bluca) [inline src/shared/seccomp-util.c]: add "ignoring"? Given it continues

**Comment by poettering** [comment]
Made all requested changes. ptal

**Comment by keszybz** [comment]
> Still I'd like to ask both bash and zsh completion to be included in PRs which add new verb or option. But, that's not requirement. Let's merge this as is.

FWIW, I agree with Yu here: a trivial addition of an option is a select&paste operation and it's just easier to do it all in one go than to deal with a separate pull request later.

**Pull request merged**